packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_data.dart

```diff
@@ -2,40 +2,88 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+import 'package:collection/collection.dart';
 import 'package:flutter/foundation.dart';
 
-import '../../../../../shared/primitives/simple_items.dart';
+import '../../../../../../devtools_app.dart';
 import '../../../shared/primitives/memory_timeline.dart';
 import '../data/primitives.dart';
 
+class _Json {
+  static const isDeviceAndroid = 'isAndroid';
+  static const timeline = 'timeline';
+  static const interval = 'interval';
+  static const isLegendVisible = 'isLegendVisible';
+}
+
 /// Chart data, that should be saved when transferred to offline data mode.
 class ChartData {
-  ChartData({required this.isDeviceAndroid});
+  ChartData({
+    required DevToolsMode mode,
+    this.isDeviceAndroid,
+    MemoryTimeline? timeline,
+    ChartInterval? interval,
+    bool? isLegendVisible,
+  }) : assert(
+          mode == DevToolsMode.connected ||
+              (mode == DevToolsMode.offlineData &&
+                  isDeviceAndroid != null &&
+                  timeline != null &&
+                  interval != null &&
+                  isLegendVisible != null),
+        ) {
+    this.timeline = timeline ?? MemoryTimeline();
+    _displayInterval =
+        ValueNotifier<ChartInterval>(interval ?? ChartInterval.theDefault);
+    _isLegendVisible = ValueNotifier<bool>(isLegendVisible ?? true);
```

Comment:
instead of making these late variables, leave them as final and then just call `.value = ...` here in the constructor body

Reply:
+1, `ValueNotifier`s should always be assigned to avoid introducing `late`, which does insert the equivalent of a null check to each access of the variable.